vnext/Shared/WinRTHttpResource.cpp

```diff
@@ -0,0 +1,296 @@
+        }
+        return;
+      } else if (header.first == "content-encoding") {
+        contentEncoding = header.second;
+      } else if (header.first == "content-length") {
+        contentLength = header.second;
+      } else if (header.first == "authorization") {
+        bool success =
+            request.Headers().TryAppendWithoutValidation(to_hstring(header.first), to_hstring(header.second));
+        if (m_onError) {
+          m_onError(requestId, "Failed to append Authorization");
+        }
+        return;
+      } else {
+        request.Headers().Append(to_hstring(header.first), to_hstring(header.second));
+      }
+    }
+
+    IHttpContent content{nullptr};
+    if (BodyData::Type::String == bodyData.Type) {
+      content = HttpStringContent{to_hstring(bodyData.Data)};
+    } else if (BodyData::Type::Base64 == bodyData.Type) {
+      auto buffer = CryptographicBuffer::DecodeFromBase64String(to_hstring(bodyData.Data));
+      content = HttpBufferContent{buffer};
+    } else if (BodyData::Type::Uri == bodyData.Type) {
+      auto file = StorageFile::GetFileFromApplicationUriAsync(Uri{to_hstring(bodyData.Data)}).get(); // TODO: async??
+      auto stream = file.OpenReadAsync().get();
+      content = HttpStreamContent{stream};
+    } else if (BodyData::Type::Form == bodyData.Type) {
+      // TODO: Add support
+    } else {
+      // BodyData::Type::Empty
+      // TODO: Error 'cause unsupported??
+    }
+
+    if (content != nullptr) {
+      // Attach content headers
+      if (contentType) {
+        content.Headers().ContentType(contentType);
+      }
+      if (!contentEncoding.empty()) {
+        if (!content.Headers().ContentEncoding().TryParseAdd(to_hstring(contentEncoding))) {
+          if (m_onError) {
+            m_onError(requestId, "Failed to parse Content-Encoding");
+          }
+          return;
+        }
+      }
+      if (!contentLength.empty()) {
+        const auto contentLengthHeader = _atoi64(contentLength.c_str()); // TODO: Check error?
+        content.Headers().ContentLength(contentLengthHeader);
+      }
+
+      request.Content(content);
+    }
+
+    PerformSendRequest(requestId, request, responseType == "text");
+  } catch (...) { // TODO: Delcare specific exception types
+    // TODO: OnRequestError
+  }
+}
+
+void WinRTHttpResource::AbortRequest(int64_t requestId) noexcept /*override*/ {
+  ResponseType request{nullptr};
+
+  {
+    scoped_lock lock{m_mutex};
+    auto iter = m_requests.find(requestId);
+    if (iter == std::end(m_requests)) {
+      return;
+    }
+    request = iter->second;
+  }
+
+  try {
+    request.Cancel();
+  } catch (hresult_error const &) {
+    // TODO: Propagate error?
+  }
+}
+
+void WinRTHttpResource::ClearCookies() noexcept /*override*/ {
+  assert(false);
+  // NOT IMPLEMENTED
+}
+
+void WinRTHttpResource::SetOnRequest(function<void(int64_t requestId)> &&handler) noexcept /*override*/ {
+  m_onRequest = std::move(handler);
```

Comment:
>m_onRequest [](http://example.com/codeflow?start=2&length=11)

These are public methods, should they also be under lock?

Reply:
Theoretically yes, but only the React module consumes this API, and only runs these methods only at resource construction/setup.